### ISSUE TITLE
Azure Pipelines not running latest release of Safari Technology Preview

### ISSUE THREAD

**Issue by jugglinmike** [web-platform-tests/wpt#17186]
[Apple released version 83 of Safari Technology Preview last Wednesday](https://webkit.org/blog/8967/release-notes-for-safari-technology-preview-83/). However, according to the results published on [wpt.fyi](https://wpt.fyi), this project continues to use version 82.

This is surprising because [the URL referenced by the Azure Pipelines configuration](https://github.com/web-platform-tests/wpt/blob/c027c41a085a874404922c638387c97df74b4ebf/tools/ci/azure/install_safari.yml#L10) currently resolves to [a script that's documented to install ~~version 63~~ version 83](https://github.com/Homebrew/homebrew-cask-versions/blob/244ffc16d5c313fdf45d2c68f34cf1a31db52f7d/Casks/safari-technology-preview.rb).

@foolip any thoughts on what's going wrong?

**Comment by Hexcles** [comment]
> This is surprising because the URL referenced by the Azure Pipelines configuration currently resolves to a script that's documented to install version 63.

Where is 63 documented?

The formula (or whatever Homebrew calls it) has been updated to point to 83 last week: https://github.com/Homebrew/homebrew-cask-versions/pull/7417

**Comment by jugglinmike** [comment]
Oh, pardon me. My original message has a typo. The text "a script that's documented to install version 63" ought to read, "a script that's documented to install version 83"

I'll update the description to avoid more confusion.

**Comment by foolip** [comment]
I think the problem is that https://github.com/Homebrew/homebrew-cask-versions/commit/55b3c016358c28322b0844041142dc3902127771 didn't update the URL for `if MacOS.version <= :high_sierra`.

https://github.com/foolip/safari-technology-preview-updater hasn't been running correctly in a while, the scheduled builds silently stopped. I'll see if getting it back and running will produce a PR fixing this.

**Comment by jugglinmike** [comment]
Good eye!

**Comment by foolip** [comment]
Also, https://developer.apple.com/safari/download/ now has packages for Mojave + Catalina. As long as our VMs are on High Sierra we just won't be able to run new Safari TP I think :(

**Comment by foolip** [comment]
Oh wait, look at this good news! https://devblogs.microsoft.com/devops/azure-pipelines-now-supports-additional-hosted-macos-versions/

**Comment by gsnedders** [comment]
Yeah, I was about to ask why we're on High Sierra. :)

**Comment by foolip** [comment]
In https://github.com/web-platform-tests/wpt/pull/18610 I tried upgrading both macOS and STP at the same time, but it's not working.

**Comment by foolip** [comment]
The problem is with the `safaridriver --enable` step. Locally, `sudo "/Applications/Safari Technology Preview.app/Contents/MacOS/safaridriver" --enable` doesn't work, but without sudo and entering password works.

I'm guessing this setting is now per-user, and that sudo sets it for the root user, perhaps.

**Comment by foolip** [comment]
`sudo -u `whoami` "/Applications/Safari Technology Preview.app/Contents/MacOS/safaridriver" --enable` also doesn't work.

@burg is there a new way to enable safaridriver for STP in CI?

**Comment by foolip** [comment]
There have been a number of PRs around upgrading macOS and STP where there's discussion that can be relevant to this issue. In order of relevance:
- [Upgrade macOS VMs to Mojave](https://github.com/web-platform-tests/wpt/pull/17219) identified WebKit bug now resolved
- [Upgrade to Safari Technology Preview 91 on macOS Mojave](https://github.com/web-platform-tests/wpt/pull/18925) is the latest attempt to upgrade, blocked on `safaridriver --enable` requiring password
- [Upgrade Safari stable runs to use macOS Mojave](https://github.com/web-platform-tests/wpt/pull/18927) concluded we shouldn't
- [Enable `safaridriver --diagnose` logs](https://github.com/web-platform-tests/wpt/pull/18444) may come in handy

Right now, this is blocked on a way to enable `safaridriver` without the user password, or by just knowing the vsts user password, which I asked @mustjab about in https://github.com/web-platform-tests/wpt/pull/18925#issuecomment-529882851.

**Comment by foolip** [comment]
I've bisected the change in behavior of `sudo /Applications/Safari\ Technology\ Preview.app/Contents/MacOS/safaridriver --enable` to https://github.com/Homebrew/homebrew-cask-versions/commit/3c75f85d8e2eacf8fbc2a6c65bfd3f29cec513a6 / https://github.com/Homebrew/homebrew-cask-versions/pull/7656, in other words it worked in STP 87 and no longer works in STP 88.

Additionally, installing 87 and enabling automation and then upgrading *doesn't* preserve the setting, so it's not an option to enable it with an older version as a workaround.

**Comment by foolip** [comment]
Ha!! https://github.com/web-platform-tests/wpt/pull/18925/commits/f31328c4c2fd102f8a39208e90e2f1f88f441284 worked around the problem by just creating ~/Library/WebDriver/com.apple.SafariTechnologyPreview.plist and skipping `safaridriver --enable`.

In case it might help another poor soul trying to work out a problem like this, this is how I worked it out on aMacbook:
- Run `sudo fs_usage -w` in one terminal
- `/Applications/Safari\ Technology\ Preview.app/Contents/MacOS/safaridriver -p 7777` in another
- `curl -d '{"capabilities":{}}' POST http://localhost:7777/session` in yet another
- Do this with the remote debugging setting enabled or disabled (via `safaridriver --enable` and password)

I couldn't spot what files `safaridriver --enable` wrote because of everything else going on, but what eventually worked was looking through the `fs_usage` logs when doing the `curl` call. That showed `/Users/foolip/Library/WebDriver/com.apple.SafariTechnologyPreview.plist` being read, and unlike a lot of other plist files being opened, it actually existed. Looks like creating that file is the important thing that `safaridriver --enable` does :D

**Comment by foolip** [comment]
However, it looks like https://bugs.webkit.org/show_bug.cgi?id=174030 is not resolved in STP 91. I'll see how many tests are improved vs. regressed with the upgrade to see if it's worth taking the regressions or not.

**Comment by foolip** [comment]
I've closed https://github.com/web-platform-tests/wpt/pull/18925 and think we should wait for STP 92, but that branch can be reused then for the `safaridriver --enable` workaround.

**Comment by burg** [comment]
Thanks for bisecting, aside from standalone repro steps this is the most helpful thing you can do towards a fix on safaridriver's end.

Manually toggling the Allow Remote Automation flag in the Develop menu did seem to have the expected effect immediately.

I was able to reproduce it not working via the following steps:

1. toggle off Enable Remote Automation in Develop menu
2. Keep Safari Technology Preview open
3. $STP/safaridriver --enable 
At this point, the Enable Remote Automation item becomes checked.

4. $STP/safaridriver -p 12345
5. New Session request
=> "SessionNotCreatedException: Message: Could not create a session: You must enable the 'Allow Remote Automation' option in Safari's Develop menu to control Safari via WebDriver."

If I quit Safari Technology Preview after step (3), then step (5) relaunches it correctly and starts an Automation session. If Safari Technology Preview is left open, it fails indefinitely until Safari Technology Preview is restarted.

So, in summary it seems like Safari Technology Preview is not correctly advertising its new automation availability status when turned on via safaridriver --enable.

This also seems to affect System Safari included macOS Catalina / Safari 13 on downlevel OS'es. The workaround in any case should be to kill the Safari after running --enable.

We are tracking this internally as rdar://problem/55267022. I am testing out a fix today.

**Comment by burg** [comment]
Interesting, that definitely sounds like a different issue. I'll investigate this as well, though it may take more time due to the repro steps. 

One thing to note is that STP 88+ does migrate the Allow Remote Automation setting from STP 87 and earlier. However, only STP can do this migration, not safaridriver. So simply launching Safari Technology Preview once and quitting should be enough to migrate an existing preference.

**Comment by burg** [comment]
> @burg from your description it sounds like that might be a different issue. Note that in CI (Azure Pipelines) things happen in this order:
> 
> * fresh VM without Safari Technology Preview starts
> * Install Safari Technology Preview from [brew cask](https://github.com/Homebrew/homebrew-cask-versions/blob/master/Casks/safari-technology-preview.rb)
> * Run `sudo /Applications/Safari\ Technology\ Preview.app/Contents/MacOS/safaridriver --enable`
> * Run `./wpt run` which will internally launch safaridriver for the first time.
> 
> In other words, neither STP nor safaridriver has never run when we try to enable remote automation.

I tried doing this and reproduced with `sudo safaridriver --enable` and didn't reproduce with `safaridriver --enable`.

**Comment by foolip** [comment]
If you mean that you weren't able to enable remote automation with sudo, then that's exactly the issue I bisected for.

**Comment by burg** [comment]
Alright. I'm tracking that issue separately, it should be easy to investigate as it does not require a clean install to reproduce.

**Comment by foolip** [comment]
Thanks! From the behavior I gatheree that the enable automation flag had been made per-user and that sudo would set it for the wrong user (root) but it sounds like this wasn't an expected change. Whatever the cause, I'm looking forward to see it resolved :)

**Comment by foolip** [comment]
STP 92 was just released, but we're still stuck because my lovely plist workaround no longer works:
https://github.com/web-platform-tests/wpt/pull/19142

**Comment by foolip** [comment]
In other words, waiting for STP 93 or if someone can find a new hack. I don't have my mac now though so can't experiment outside of CI.

**Comment by foolip** [comment]
So I tried to find a workaround for STP 92 similar to what I did for https://github.com/web-platform-tests/wpt/issues/17186#issuecomment-530306239, but it seems like the problem is now even more serious: I can't enable automation in any way as a real user on a local machine.

I've tried toggling the 'Allow Remote Automation' in the menu and running `safaridriver --enable` with and without sudo, for both Safari stable and STP (just in case they'd share some state) and still  `curl -d '{"capabilities":{}}' http://localhost:7777/session` fails with "Could not create a session: You must enable the 'Allow Remote Automation' option in Safari's Develop menu to control Safari via WebDriver."

It is still possible to control Safari stable, it's only STP that's affected.

@burg can you file an issue for this?

A smoke test in CI that previously has no STP installed using something minimal like https://github.com/foolip/safari-ci-starter-kit before STP releases would be great. I'm no longer maintaining that, but it would be excellent if someone did maintain it so that everyone can use Safari in public CI systems with minimal effort :)

**Comment by burg** [comment]
> On Sep 20, 2019, at 12:56 AM, Philip Jägenstedt <notifications@github.com> wrote:
> 
> So I tried to find a workaround for STP 92 similar to what I did for #17186 (comment) <https://github.com/web-platform-tests/wpt/issues/17186#issuecomment-530306239>, but it seems like the problem is now even more serious: I can't enable automation in any way as a real user on a local machine.
> 
> I've tried toggling the 'Allow Remote Automation' in the menu and running safaridriver --enable with and without sudo, for both Safari stable and STP (just in case they'd share some state) and still curl -d '{"capabilities":{}}' http://localhost:7777/session fails with "Could not create a session: You must enable the 'Allow Remote Automation' option in Safari's Develop menu to control Safari via WebDriver."
> 
> It is still possible to control Safari stable, it's only STP that's affected.
> 
> @burg <https://github.com/burg> can you file an issue for this?
> 
I have fixed bugs for the normal and sudo case. The normal case, where the preference change does not sync immediately, will be fixed in STP93. The sudo case, which simply does not work at all, will be fixed in STP94.

**Comment by foolip** [comment]
Thanks @burg! For STP93, do you think some workaround by writing the appropriate files to disk will be possible, or did the fixes you made not have anything to do with why that stopped working between 91 and 92? Waiting for 94 wouldn't be a huge deal, but the sooner we can have a full run the sooner we'll know if there are more issues affecting results that also need to be reported.

**Comment by foolip** [comment]
STP 93 was just released, attempting upgrade now in https://github.com/web-platform-tests/wpt/pull/19482.

**Comment by foolip** [comment]
STP 94 was just released and it finally worked. https://github.com/web-platform-tests/wpt/pull/19750 has been merged and we should hopefully have a master run with STP 94 in ~3 hours.

**Comment by foolip** [comment]
Success! We now have results for STP 94 aligned with experimental Chrome and Firefox:
https://wpt.fyi/results/?run_id=315080029&run_id=315020045&run_id=317160019

We were stuck for a long time, let's hope STP 95 is a smooth upgrade :)

**Comment by clopez** [comment]
Cool! :)

**Comment by foolip** [comment]
For anyone following along here, you may be happy to know that the upgrade from STP 94 to 95 was fast and smooth. The only point of human intervention there was in approving/merging https://github.com/Homebrew/homebrew-cask-versions/pull/8076.

**Comment by akshaygit09** [comment]
Can anyone update how to resolve this ?
![IMG_20230608_204214](https://github.com/web-platform-tests/wpt/assets/52966977/1e1c0a99-aa64-4a4c-9ba9-7e3eba7a463f)
I am getting the exception System.Reflection.TargetInvocationException even after using "sudo safaridriver --enable" and MacOS agent
